# HG changeset patch
# User Stefan Kuhn <wuodan@hispeed.ch>
# Date 1454170433 0
#      Sat Jan 30 16:13:53 2016 +0000
# Node ID 393d3797ab71499da1b4ec5b59156c9f71c7b9fe
# Parent  57617214b7b867f3e8d3600089657cdfbcee36d9
[PATCH] feature: progress bar
Show a visual progress bar on slow operations

Credits:
 * Rocco Rutte <pdmef@gmx.net>
 * Vincent Lefevre <vincent@vinc17.org>
 * Stefan Kuhn <wuodan@hispeed.ch>
 * Karel Zak <kzak@redhat.com>
 * Richard Russon <rich@flatcap.org>
---
 PATCHES             |   1 +
 color.c             |   1 +
 curs_lib.c          |  65 +++++++++++++++++++++++++---
 doc/manual.xml.head | 120 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 doc/muttrc.progress |  10 +++++
 doc/vimrc.progress  |   5 +++
 mutt_curses.h       |   1 +
 7 files changed, 247 insertions(+), 5 deletions(-)
 create mode 100644 doc/muttrc.progress
 create mode 100644 doc/vimrc.progress

diff -r 57617214b7b8 -r 393d3797ab71 PATCHES
--- a/PATCHES	Thu Apr 21 02:32:18 2016 +0100
+++ b/PATCHES	Sat Jan 30 16:13:53 2016 +0000
@@ -1,3 +1,4 @@
+patch-progress-neomutt
 patch-sidebar-on-right-neomutt
 patch-cond-date-neomutt
 patch-nested-if-neomutt
diff -r 57617214b7b8 -r 393d3797ab71 color.c
--- a/color.c	Thu Apr 21 02:32:18 2016 +0100
+++ b/color.c	Sat Jan 30 16:13:53 2016 +0000
@@ -94,6 +94,7 @@ static const struct mapping_t Fields[] =
   { "bold",		MT_COLOR_BOLD },
   { "underline",	MT_COLOR_UNDERLINE },
   { "index",		MT_COLOR_INDEX },
+  { "progress",		MT_COLOR_PROGRESS },
   { "prompt",		MT_COLOR_PROMPT },
 #ifdef USE_SIDEBAR
   { "sidebar_divider",	MT_COLOR_DIVIDER },
diff -r 57617214b7b8 -r 393d3797ab71 curs_lib.c
--- a/curs_lib.c	Thu Apr 21 02:32:18 2016 +0100
+++ b/curs_lib.c	Sat Jan 30 16:13:53 2016 +0000
@@ -478,6 +478,61 @@ void mutt_progress_init (progress_t* pro
   mutt_progress_update (progress, 0, 0);
 }
 
+/**
+ * message_bar - Draw a colourful progress bar
+ * @percent: %age complete
+ * @fmt:     printf(1)-like formatting string
+ * @...:     Arguments to formatting string
+ *
+ */
+static void
+message_bar (int percent, const char *fmt, ...)
+{
+	va_list ap;
+	char buf[STRING], buf2[STRING];
+	int w = percent * COLS / 100;
+	size_t l;
+
+	va_start (ap, fmt);
+	vsnprintf (buf, sizeof (buf), fmt, ap);
+	l = mutt_strwidth (buf);
+	va_end (ap);
+
+	mutt_format_string (buf2, sizeof (buf2), 0, COLS-2, FMT_LEFT, 0, buf, sizeof (buf), 0);
+
+	move (LINES - 1, 0);
+
+	if (ColorDefs[MT_COLOR_PROGRESS] == 0) {
+		addstr (buf2);
+	} else {
+		if (l < w) {
+			/* The string fits within the colour bar */
+			SETCOLOR(MT_COLOR_PROGRESS);
+			addstr (buf2);
+			w -= l;
+			while (w--) {
+				addch (' ');
+			}
+			SETCOLOR(MT_COLOR_NORMAL);
+		} else {
+			/* The string is too long for the colour bar */
+			char ch;
+			int off = mutt_wstr_trunc (buf2, sizeof (buf2), w, NULL);
+
+			ch = buf2[off];
+			buf2[off] = 0;
+			SETCOLOR(MT_COLOR_PROGRESS);
+			addstr (buf2);
+			buf2[off] = ch;
+			SETCOLOR(MT_COLOR_NORMAL);
+			addstr (&buf2[off]);
+		}
+	}
+
+	clrtoeol();
+	mutt_refresh();
+}
+
 void mutt_progress_update (progress_t* progress, long pos, int percent)
 {
   char posstr[SHORT_STRING];
@@ -528,16 +583,16 @@ void mutt_progress_update (progress_t* p
 
     if (progress->size > 0)
     {
-      mutt_message ("%s %s/%s (%d%%)", progress->msg, posstr, progress->sizestr,
-		    percent > 0 ? percent :
-		   	(int) (100.0 * (double) progress->pos / progress->size));
+      message_bar ((percent > 0) ? percent : (int) (100.0 * (double) progress->pos / progress->size),
+        "%s %s/%s (%d%%)", progress->msg, posstr, progress->sizestr,
+        (percent > 0) ? percent : (int) (100.0 * (double) progress->pos / progress->size));
     }
     else
     {
       if (percent > 0)
-	mutt_message ("%s %s (%d%%)", progress->msg, posstr, percent);
+        message_bar (percent, "%s %s (%d%%)", progress->msg, posstr, percent);
       else
-	mutt_message ("%s %s", progress->msg, posstr);
+        mutt_message ("%s %s", progress->msg, posstr);
     }
   }
 
diff -r 57617214b7b8 -r 393d3797ab71 doc/manual.xml.head
--- a/doc/manual.xml.head	Thu Apr 21 02:32:18 2016 +0100
+++ b/doc/manual.xml.head	Sat Jan 30 16:13:53 2016 +0000
@@ -2749,6 +2749,7 @@ When set, color is applied only to the e
 <listitem><para>markers (the <quote>+</quote> markers at the beginning of wrapped lines in the pager)</para></listitem>
 <listitem><para>message (informational messages)</para></listitem>
 <listitem><para>normal</para></listitem>
+<listitem><para><link linkend="progress">progress</link> (visual progress bar)</para></listitem>
 <listitem><para>prompt</para></listitem>
 <listitem><para>quoted (text matching <link linkend="quote-regexp">$quote_regexp</link> in the body of a message)</para></listitem>
 <listitem><para>quoted1, quoted2, ..., quoted<emphasis>N</emphasis> (higher levels of quoting)</para></listitem>
@@ -8361,6 +8362,75 @@ please have a look at the mixmaster docu
 
 </sect1>
 
+<sect1 id="progress">
+  <title>Progress Bar Patch</title>
+  <subtitle>Show a visual progress bar on slow operations</subtitle>
+
+  <sect2 id="progress-intro">
+    <title>Introduction</title>
+
+        <para>
+    The <quote>progress</quote> patch shows a visual progress bar on slow
+    tasks, such as indexing a large folder over the net.
+        </para>
+  </sect2>
+
+  <sect2 id="progress-colors">
+    <title>Colors</title>
+    <table id="table-progress-colors">
+      <title>Progress Colors</title>
+      <tgroup cols="3">
+        <thead>
+          <row>
+            <entry>Name</entry>
+            <entry>Default Color</entry>
+            <entry>Description</entry>
+          </row>
+        </thead>
+        <tbody>
+          <row>
+            <entry><literal>progress</literal></entry>
+            <entry>default</entry>
+            <entry>Visual progress bar</entry>
+          </row>
+        </tbody>
+      </tgroup>
+    </table>
+  </sect2>
+
+  <sect2 id="progress-muttrc">
+    <title>Muttrc</title>
+<screen>
+<emphasis role="comment"># Example Mutt config file for the 'progress' patch.
+ 
+# The 'progress' patch provides clear visual feedback for
+# slow tasks, such as indexing a large folder over the net.
+ 
+# Set the color of the progress bar
+# White text on a red background</emphasis>
+color progress white red
+ 
+<emphasis role="comment"># vim: syntax=muttrc</emphasis>
+</screen>
+  </sect2>
+
+  <sect2 id="progress-known-bugs">
+    <title>Known Bugs</title>
+    <para>None</para>
+  </sect2>
+
+  <sect2 id="progress-credits">
+    <title>Credits</title>
+    <itemizedlist>
+    <listitem><para>Rocco Rutte <email>pdmef@gmx.net</email></para></listitem>
+    <listitem><para>Vincent Lefevre <email>vincent@vinc17.org</email></para></listitem>
+    <listitem><para>Stefan Kuhn <email>wuodan@hispeed.ch</email></para></listitem>
+    <listitem><para>Karel Zak <email>kzak@redhat.com</email></para></listitem>
+    <listitem><para>Richard Russon <email>rich@flatcap.org</email></para></listitem>
+    </itemizedlist>
+  </sect2>
+</sect1>
+
 <sect1 id="sidebar">
   <title>Sidebar</title>
   <subtitle>Overview of mailboxes</subtitle>
diff -r 57617214b7b8 -r 393d3797ab71 doc/muttrc.progress
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/doc/muttrc.progress	Sat Jan 30 16:13:53 2016 +0000
@@ -0,0 +1,10 @@
+# Example Mutt config file for the 'progress' patch.
+
+# The 'progress' patch provides clear visual feedback for
+# slow tasks, such as indexing a large folder over the net.
+
+# Set the color of the progress bar
+# White text on a red background
+color progress white red
+
+# vim: syntax=muttrc
diff -r 57617214b7b8 -r 393d3797ab71 doc/vimrc.progress
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/doc/vimrc.progress	Sat Jan 30 16:13:53 2016 +0000
@@ -0,0 +1,5 @@
+" Vim syntax file for the mutt progress patch
+
+syntax keyword muttrcColorField contained progress
+
+" vim: syntax=vim
diff -r 57617214b7b8 -r 393d3797ab71 mutt_curses.h
--- a/mutt_curses.h	Thu Apr 21 02:32:18 2016 +0100
+++ b/mutt_curses.h	Sat Jan 30 16:13:53 2016 +0000
@@ -114,6 +114,7 @@ enum
   MT_COLOR_UNDERLINE,
   MT_COLOR_INDEX,
   MT_COLOR_PROMPT,
+  MT_COLOR_PROGRESS,
 #ifdef USE_SIDEBAR
   MT_COLOR_DIVIDER,
   MT_COLOR_FLAGGED,
